### PR TITLE
Deletion of experimental objects when Namespace is removed.

### LINKED ISSUE #12735: HorizontalPodAutoScaler object is not removed when Namespace is deleted
This is a reminder for @jszczepkowski  to make sure that it gets cleaned up.

### PR DESCRIPTION
**Pull request by jszczepkowski**
Implemented removal of Deployments, Daemons & HorizontalPodAutoscalers when Namespace is removed. Added unittest. Fixes #12735.

### CODE REVIEW

**Comment by jszczepkowski** [comment]
CC @vishh @nikhiljindal @mwielgus @derekwaynecarr

**Comment by derekwaynecarr** [comment]
My understanding is we are not wanting to pursue this pattern of having controllers handle multiple clients.  Instead, we want a layer in front of client.Interface that lets you get to Core or Experiemental objects.

See for details: https://github.com/kubernetes/kubernetes/issues/13388

I think we should block changing NamespaceController until we resolve client management pattern to let you access both Core and Experimental object from the same client resource level.

**Comment by jszczepkowski** [comment]
@derekwaynecarr Modification in NamespaceController are really simple. They can be trivially refactored later. Modification in testclient shouldn't be really affected by merging clients.

In my opinion, it is better to implement missing functionality now, even if some refactoring may be needed in future. (This PR is fixing P1 issue)! We don't want to block implementation of horizontal pod autoscaler (it is using to clients for now) as we don't have that much time left for version 1.1.

Removing objects when namespace is removed is badly needed for e2e tests (work on them is already in progress)

I really think that potential future refactoring of experimental client shouldn't block this PR.

**Comment by derekwaynecarr** [comment]
@jszczepkowski - as long as its fixed afterward and does not become the de-facto standard pattern I am fine.

fyi @timothysc @lavalamp - this pr is putting a band-aid on our client problem with core/experimental but is fixing up the namespace_controller to delete experimental resources.

**Comment by wojtek-t** [comment]
cc @kubernetes/goog-csi

**Comment by lavalamp** [comment]
On Thu, Sep 3, 2015 at 10:36 AM, Derek Carr <notifications@github.com>
wrote:

> @jszczepkowski <https://github.com/jszczepkowski> - as long as its fixed
> afterward and does not become the de-facto standard pattern I am fine.
>
> fyi @timothysc <https://github.com/timothysc> @lavalamp
> <https://github.com/lavalamp> - this pr is putting a band-aid on our
> client problem with core/experimental but is fixing up the
> namespace_controller to delete experimental resources.
>
Acknowledged.

**Comment by derekwaynecarr** [comment]
@k8s-bot test again

File: pkg/controller/namespace/namespace_controller.go [hunk 1]
```diff
@@ -189,12 +189,26 @@ func deleteAllContent(kubeClient client.Interface, namespace string, before util
 	if err != nil {
 		return estimate, err
 	}
-
+	// If experimental client is set, delete all experimental resrources for the namespace.
```
Comment (derekwaynecarr) [inline pkg/controller/namespace/namespace_controller.go]: nit: delete all experimental resources

**Comment by timothysc** [comment]
imho #13523, then this.

**Comment by jszczepkowski** [comment]
> imho #13523, then this.

Sure, It can wait for #13523

File: pkg/controller/namespace/namespace_controller.go (hunk 1 above)
Comment (jszczepkowski) [inline pkg/controller/namespace/namespace_controller.go]: Done

**Comment by jszczepkowski** [comment]
@derekwaynecarr I've patched my PR after #13523 was merged. It is ready for the review.

**Comment by timothysc** [comment]
@rrati - PTAL similar issue.

File: cmd/kube-controller-manager/app/controllermanager.go [hunk 2]
```diff
@@ -241,9 +241,14 @@ func (s *CMServer) Run(_ []string) error {
 	resourceQuotaController := resourcequotacontroller.NewResourceQuotaController(kubeClient)
 	resourceQuotaController.Run(s.ResourceQuotaSyncPeriod)
 
-	namespaceController := namespacecontroller.NewNamespaceController(kubeClient, s.NamespaceSyncPeriod)
+	namespaceController := namespacecontroller.NewNamespaceController(kubeClient, s.EnableHorizontalPodAutoscaler, s.NamespaceSyncPeriod)
 	namespaceController.Run()
```
Comment (rrati) [inline cmd/kube-controller-manager/app/controllermanager.go]: Can we make EnableHorizontalPodAutoscaler more generic?  There may be other experimental resources that will need to be cleaned up when a namespace is deleted.

Comment (jszczepkowski) [inline cmd/kube-controller-manager/app/controllermanager.go]: I'm not sure if we want it. There may be many other experimental controllers (deployments, deamons, ...). I think we should have a separate flag for enabling each of them, and take alternative of them here.

Comment (derekwaynecarr) [inline cmd/kube-controller-manager/app/controllermanager.go]: Maybe just a separate line so folks that want to put flags behind their feature its easier to see where they should add the code in the future.

```
// An OR of all flags to enable/disable experimental features
experimentalMode := s.EnableHorizontalPodAutoscaler
namespaceController := ....
```

File: pkg/controller/namespace/namespace_controller_test.go [hunk 3]
```diff
@@ -85,7 +85,7 @@ func TestSyncNamespaceThatIsTerminating(t *testing.T) {
 			Phase: api.NamespaceTerminating,
 		},
 	}
-	err := syncNamespace(mockClient, testNamespace)
+	err := syncNamespace(mockClient, true, testNamespace)
```
Comment (derekwaynecarr) [inline pkg/controller/namespace/namespace_controller_test.go]: Can you do a test with this false and verify you get a reduced action set?

**Comment by derekwaynecarr** [comment]
Needs rebase.  Few minor comments, and request for additional test then lgtm.

File: cmd/kube-controller-manager/app/controllermanager.go (hunk 2 above)
Comment (timothysc) [inline cmd/kube-controller-manager/app/controllermanager.go]: +1.

**Comment by timothysc** [comment]
Waiting patiently, some other changes are pending on this one.

File: cmd/kube-controller-manager/app/controllermanager.go (hunk 2 above)
Comment (jszczepkowski) [inline cmd/kube-controller-manager/app/controllermanager.go]: Done.

File: pkg/controller/namespace/namespace_controller_test.go (hunk 3 above)
Comment (jszczepkowski) [inline pkg/controller/namespace/namespace_controller_test.go]: Done.

**Comment by jszczepkowski** [comment]
@derekwaynecarr I've applied all the comments and re-based. PTAL

**Comment by jszczepkowski** [comment]
@k8s-bot test again

**Comment by derekwaynecarr** [comment]
@k8s-bot - test again

**Comment by derekwaynecarr** [comment]
@k8s-bot test this

**Comment by derekwaynecarr** [comment]
LGTM

**Comment by jszczepkowski** [comment]
Part of #12087.

**Pull request merged**